002-es-modules.md

```diff
@@ -0,0 +1,540 @@
+
+#### 3.3.3. Shipping both ESM and NCJS
+
+When a `package.json` main is encountered, file extension searches are used to
+provide a means to ship both ESM and NCJS variants of packages. If we have two
+entry points `index.mjs` and `index.js` setting `"main":"./index"` in
+`package.json` will make Node pick up either, depending on what is supported.
+
+##### 3.3.3.1. Excluding main
+
+Since `main` in `package.json` is entirely optional even inside of npm
+packages, some people may prefer to exclude main entirely in the case of using
+`./index` as that is still in the Node module search algorithm.
+
+### 3.4. ESM Evaluation
+
+#### 3.4.1. Environment Variables
+
+ESM will not be bootstrapped with magic variables and will await upcoming specifications in order to provide such behaviors in a standard way. As such, the following variables are changed:
+
+| Variable | Exists | Value |
+| ---- | ---- | --- |
+| this | y | [`undefined`](https://tc39.github.io/ecma262/#sec-module-environment-records-getthisbinding) |
+| arguments | n | |
+| require | n | |
+| module | n | |
+| exports | n | |
+| __filename | n | |
+| __dirname | n | |
+
+Like normal scoping rules, if a variable does not exist in a scope, the outer scope is used to find the variable. Since ESM are always strict, errors may be thrown upon trying to use variables that do not exist globally when using ESM.
+
+##### 3.4.1.1. Workaround
+
+Although heavily advised against, you can have a NM module sibling for your ESM that can export these things:
+
+```js
+// expose.js
+module.exports = {__dirname};
+```
+
+```js
+// use.mjs
+import expose from './expose.js';
+const {__dirname} = expose;
+```
+
+#### 3.4.2. Timing
+
+When loading ESM or using any form of `import`, the stack will unwind prior to performing any resolution or evaluation.
+
+##### 3.4.2.1. Example
+
+```javascript
+// entry.js
+require('one.mjs');
+console.log('two');
+```
+
+```javascript
+// one.mjs
+console.log('one');
+```
+
+```sh
+> node entry.js
+two
+one
+```
+
+### 3.5. Cross Module System Communication
+
+ESM and NCJS differ in many ways, as such they need a well defined means of converting between the two module types and systems.
+
+### 3.5.1. NCJS to ESM
+
+After *any* NCJS finishes evaluation, it will be placed into the same cache as ESM.
+The value of what is placed in the cache will reflect a single default export pointing to the value of `module.exports` at the time evaluation ended.
+
+Essentially after any NCJS completes evaluation:
+
+1. if there was an error, place the error in the ESM cache and return
+2. let `export` be the value of `module.exports`
+3. if there was an error, place the error in the ESM cache and return
+4. create an ESM with `{default:export}` as its namespace
+5. place the ESM in the ESM cache
+
+### 3.5.2. ESM to NCJS
+
+ESM are *never* placed in the NCJS cache (`require.cache`).
+
+If `require` resolves to an ESM source text, the following steps are taken.
```

Comment:
As it is, this is just going to require people (and transpilers) to wrap every single `require` call in `Promise.resolve`, or at least `Promise.all`.

Updating modules to support ESM will also be a major breaking change, as it will change `require`s to return a completely different data type, and will require major refactoring of all consumers of a module, unless they all preemptively use `Promise.resolve(require)`.

```js
import foo from 'commonjs/bar'
```

would have to transpile to:

```js
Promise.all([require('commonjs/bar')])
.then(([foo, es]) => {})
```

even if no ESM imports are involved, as not doing so would prevent ever migrating `commonjs/bar` to ESM without breaking everyone.